### PR TITLE
Support `[Slice]` in manage_unit and manage_dropin

### LINKED ISSUE #417: Add possibiliry to setup limits for user sessions
In service-limits.pp I see:

https://github.com/voxpupuli/puppet-systemd/blob/bcec820dfc7188f75b579f4a67e14ea1b71861e0/manifests/service_limits.pp#L46-L48

But I need to set limits for user slices, i.e. "user-.slice.d".

Please ad support for user slice limits

### PR DESCRIPTION
**Pull request by traylenator**
#### Pull Request (PR) description
For example the following is now possible:

```puppet
systemd::manage_dropin { 'userlimits.conf':
  unit  => 'user-.slice',
  slice_entry => {
    MemoryMax        => '10G',
    MemoryAccounting => true,
  }
}
```

The directives for `[Slice]` are typically a subset of `[Service]`.

* https://www.freedesktop.org/software/systemd/man/latest/systemd.slice.html
* https://www.freedesktop.org/software/systemd/man/latest/systemd.resource-control.html

Adds new types:
* Systemd::Unit::Amount
* Systemd::Unit::Percent
* Systemd::Unit::AmountOrPercent

#### This Pull Request (PR) fixes the following issues
* Fixes #417

### CODE REVIEW

File: REFERENCE.md [hunk 1]
```diff
@@ -2495,6 +2526,48 @@ Possible strings for ExecStart, ExecStartPrep, ...
...
+  * https://www.freedesktop.org/software/systemd/man/systemd.slice.html
+  * https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html
+
+Alias of
+
+```puppet
+Struct[{
+    Optional['CPUAccounting']       => Boolean,
+    Optional['CPUQuota']            => Pattern['^([1-9][0-9]*)%$'],
+    Optional['CPUShares']           => Integer[2,262144],
```
Comment (TheMeier) [inline REFERENCE.md:2543]: `CPUShares` and a couple of others are deprecated.
https://www.freedesktop.org/software/systemd/man/latest/systemd.resource-control.html#systemd%20252

Of cause there is still systems which do not provide cgroups v2 yet, but for a new feature I wonder if it is worth implementing support for those.

File: REFERENCE.md [hunk 2]
```diff
@@ -2495,6 +2526,48 @@ Possible strings for ExecStart, ExecStartPrep, ...
...
+    Optional['CPUWeight']           => Variant[Enum['idle'],Integer[1,10000]],
+    Optional['Delegate']            => Boolean,
+    Optional['DeviceAllow']         => String[1],
```
Comment (TheMeier) [inline REFERENCE.md:2546]: Maybe make this more strict (dedicated type?)

> The device node specifier is either a path to a device node in the file system, starting with /dev/, or a string starting with either "char-" or "block-"

File: REFERENCE.md [hunk 3]
```diff
@@ -2495,6 +2526,48 @@ Possible strings for ExecStart, ExecStartPrep, ...
...
+    Optional['IOWriteBandwidthMax'] => Array[Hash[Stdlib::Absolutepath, Pattern['^(\d+(K|M|G|T)?)$'], 1, 1]],
+    Optional['IOWriteIOPSMax']      => Array[Hash[Stdlib::Absolutepath, Pattern['^(\d+(K|M|G|T)?)$'], 1, 1]],
+    Optional['IPAccounting']        => Boolean,
+    Optional['MemoryAccounting']    => Boolean,
+    Optional['MemoryHigh']          => Pattern['\A(infinity|\d+(K|M|G|T|%)?(:\d+(K|M|G|T|%)?)?)\z'],
+    Optional['MemoryLimit']         => Pattern['\A(infinity|\d+(K|M|G|T|%)?(:\d+(K|M|G|T|%)?)?)\z'], # dep'd in systemd
+    Optional['MemoryLow']           => Pattern['\A(infinity|\d+(K|M|G|T|%)?(:\d+(K|M|G|T|%)?)?)\z'],
+    Optional['MemoryMax']           => Pattern['\A(infinity|\d+(K|M|G|T|%)?(:\d+(K|M|G|T|%)?)?)\z'],
+    Optional['MemoryMin']           => Pattern['\A(infinity|\d+(K|M|G|T|%)?(:\d+(K|M|G|T|%)?)?)\z'],
+    Optional['MemorySwapMax']       => Pattern['\A(infinity|\d+(K|M|G|T|%)?(:\d+(K|M|G|T|%)?)?)\z'],
```
Comment (TheMeier) [inline REFERENCE.md:2562]: DRY. Why not add a dedicated type for `Pattern['\A(infinity|\d+(K|M|G|T|%)?(:\d+(K|M|G|T|%)?)?)\z']`

**Review by TheMeier — COMMENTED** [review]
There is also a couple more parameters in the latest systemd that are missing:

- `StartupCPUWeight`
- `AllowedCPUs`
-  `StartupAllowedCPUs`
- `StartupMemoryLow`
- `DefaultStartupMemoryLow`
- `StartupMemoryLow`
- `DefaultStartupMemoryLow`
- `StartupMemoryHigh`
- `StartupMemoryMax`
- `StartupMemorySwapMax`
- `MemoryZSwapMax`
- `StartupMemoryZSwapMax`
- `AllowedMemoryNodes`
- `StartupAllowedMemoryNodes`

Disclaimer: I do not have experience with managing slices. 
I wonder what the slice name `user-.slice` means. In the documentation I did only find the root slice `-.slice` 2nd level slices `foo.slice` and subsequent slices `foo-bar.slice`. So I assume `user-.slice` is equivalent to `user.slice`

File: REFERENCE.md (hunk 3 above)
Comment (ondrejv2) [reply REFERENCE.md:2562]: `user-.slice` acts as template for all user slices

File: REFERENCE.md (hunk 1 above)
Comment (traylenator) [reply REFERENCE.md:2543]: > `CPUShares` and a couple of others are deprecated.
> https://www.freedesktop.org/software/systemd/man/latest/systemd.resource-control.html#systemd%20252
> 
> Of cause there is still systems which do not provide cgroups v2 yet, but for a new feature I wonder if it is worth implementing support for those.

Think we need the  the older ones, we are still using them for sure on CentOS 7 and even EL8 is cgroups v1 by default.

**Comment by traylenator** [comment]
> There is also a couple more parameters in the latest systemd that are missing:
> 
> * `StartupCPUWeight`
> * `AllowedCPUs`
> * `StartupAllowedCPUs`
> * `StartupMemoryLow`
> * `DefaultStartupMemoryLow`
> * `StartupMemoryLow`
> * `DefaultStartupMemoryLow`
> * `StartupMemoryHigh`
> * `StartupMemoryMax`
> * `StartupMemorySwapMax`
> * `MemoryZSwapMax`
> * `StartupMemoryZSwapMax`
> * `AllowedMemoryNodes`
> * `StartupAllowedMemoryNodes`
> 
I don't want to add everything. 

Get a good set in and things like `StartupMemoryZSwapMax` are super obscure in my opinion. Let someone who uses
them add them or I'll happy add later.

**Review by TheMeier — APPROVED** [review]

**Pull request merged**